src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -63,8 +63,36 @@ void GcsNodeManager::HandleRegisterNode(rpc::RegisterNodeRequest request,
     AddNode(std::make_shared<rpc::GcsNodeInfo>(request.node_info()));
     GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
   };
-  RAY_CHECK_OK(
-      gcs_table_storage_->NodeTable().Put(node_id, request.node_info(), on_done));
+  if (request.node_info().is_head_node()) {
+    // mark all old head nodes as dead if exists:
+    // 1. should never happen when HA is not used
+    // 2. happens when a new head node is started
+
+    std::vector<NodeID> head_nodes;
+    for (auto &node : alive_nodes_) {
+      if (node.second->is_head_node()) {
+        head_nodes.push_back(node.first);
+      }
+    }
+    std::function<void(int)> kill_node_chain_callback =
+        [this, &kill_node_chain_callback, head_nodes, on_done, request, node_id](
+            const auto &node_idx) {
+          if (node_idx < static_cast<int>(head_nodes.size())) {
```

Comment:
hmm this recursion logic seems unnecessary. can you just iterate head_nodes instead? I don't think the callback is also necessary? 

```
assert len(head_nodes) == 1
OnNodeFailure(head_nodes[0])
Put(new_head)
```

Reply:
we do callback since `OnNodeFailure` is async (has KV put as well). So we want to ensure that `OnNodeFailure` finish before putting the new head node by make it sync

Reply:
it will be marked as dead as the source of truth is in memory (alive_nodes_). 

We don't need to worry about ordering as redis update ordering is guaranteed. So failure update will happen before new node update happens to Redis (meaning we don't need to worry about GCS failure case). 

cc @jjyao